### PR TITLE
Fix a couple of JsonConstructor bugs & clean up JsonParameterInfo implementation.

### LINKED ISSUE #60082: Default value for optional ctor params should be used when matching prop is ignored
Repro:

```cs
public class Point
{
    public int X { get; }

    [JsonIgnore]
    public int Y { get; }

    public Point(int x, int y = 5) => (X, Y) = (x, y);
}

string json = "{}";
var point = JsonSerializer.Deserialize<Point>(json);
Console.WriteLine(point.Y);
// Expected: 5
// Actual: 0
```

The behavior should be consistent with then the matching prop is not ignored:

```cs
public class Point
{
    public int X { get; }

    public int Y { get; }

    public Point(int x, int y = 5) => (X, Y) = (x, y);
}

string json = "{}";
var point = JsonSerializer.Deserialize<Point>(json);
Console.WriteLine(point.Y);
// Expected: 5
// Actual: 5
```

### LINKED ISSUE #68647: Fix JSON issue (InvalidCastException) with ctor nullable enum ctor params having default values
### Description

When serializing JSON using System.Text.Json.JsonSerializer an object that has a constructor with some parameter that is nullable, but has non-null default value, JsonSerializer crashes.

The error seems to be that JsonTypeInfo sees that there is a constructor that has parameters. Then it goes to 
System.Text.Json.Serialization.Metadata.JsonParameterInfo<T>.InitializeDefaultValue that sees default value as int and tries to cast it as nullable enum type that fails.

### Reproduction Steps

```c#
var ss = System.Text.Json.JsonSerializer.Serialize(new MyClass(), options);

public sealed class MyClass
{
    public MyClass(MyEnum? value = MyEnum.Value)
    {
        this.Value = value;
    }

    public MyEnum? Value { get; }
}

public enum MyEnum
{
    Value = 1,
}
```

### Expected behavior

Succesful serialization into following string: `{"Value":1}`

### Actual behavior

Crashed with following exception:
InvalidCastException
"Unable to cast object of type 'System.Int32' to type 'System.Nullable`1[MyEnum]'."
```
   at System.Text.Json.Serialization.Metadata.JsonParameterInfo`1.InitializeDefaultValue(JsonPropertyInfo matchingProperty)
   at System.Text.Json.Serialization.Metadata.JsonParameterInfo`1.Initialize(JsonParameterInfoValues parameterInfo, JsonPropertyInfo matchingProperty, JsonSerializerOptions options)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo.InitializeConstructorParameters(JsonParameterInfoValues[] jsonParameters, Boolean sourceGenMode)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo..ctor(Type type, JsonConverter converter, Type runtimeType, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializerOptions.<InitializeForReflectionSerializer>g__CreateJsonTypeInfo|112_0(Type type, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializerOptions.GetClassFromContextOrCreate(Type type)
   at System.Text.Json.JsonSerializerOptions.GetOrAddClass(Type type)
   at System.Text.Json.JsonSerializer.GetTypeInfo(JsonSerializerOptions options, Type runtimeType)
   at System.Text.Json.JsonSerializer.Serialize[TValue](TValue value, JsonSerializerOptions options)
   at JsonCrashTest.Program.Main() in Program.cs:line 15
```

### Regression?

_No response_

### Known Workarounds

Workaround seems to have constructor parameter names not to match property names.

### Configuration

Using .NET 6.0.3 on Windows 10 x64

### Other information

_No response_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix #60082. Fix #68647.

Simplifies the `JsonParameterInfo` implementation using conventions already used in `JsonTypeInfo` and `JsonPropertyInfo`.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs [hunk 1]
```diff
@@ -92,31 +90,23 @@ protected override void InitializeConstructorArgumentCaches(ref ReadStack state,
             for (int i = 0; i < typeInfo.ParameterCount; i++)
             {
                 JsonParameterInfo parameterInfo = cache[i].Value;
-
-                // We can afford not to set default values for ctor arguments when we should't deserialize because the
-                // type parameters of the `Arguments` type provide default semantics that work well with value types.
-                if (parameterInfo.ShouldDeserialize)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs:98]: This check is incorrect -- we need to populate the default constructor values even for ignored properties. This is changed to be consistent with the large constructor converter.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs [hunk 2]
```diff
@@ -11,108 +11,43 @@ namespace System.Text.Json.Serialization.Metadata
...
-        public Type PropertyType { get; set; } = null!;
+        public Type ParameterType { get; }
 
-        public bool ShouldDeserialize { get; private set; }
+        public bool ShouldDeserialize { get; }
 
-        public JsonPropertyInfo MatchingProperty { get; private set; } = null!;
+        public JsonPropertyInfo MatchingProperty { get; }
 
-        public virtual void Initialize(JsonParameterInfoValues parameterInfo, JsonPropertyInfo matchingProperty, JsonSerializerOptions options)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs:58]: Replaces `Initialize` methods with constructors -- makes it easier to guarantee non-nullability of properties.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs [hunk 3]
```diff
@@ -11,108 +11,43 @@ namespace System.Text.Json.Serialization.Metadata
...
-            jsonParameterInfo.PropertyType = matchingProperty.PropertyType;
-            jsonParameterInfo.NameAsUtf8Bytes = matchingProperty.NameAsUtf8Bytes!;
-
-            // TODO: https://github.com/dotnet/runtime/issues/60082.
-            // Default value initialization for params mapping to ignored properties doesn't
-            // account for the default value of optional parameters. This should be fixed.
-
-            if (sourceGenMode)
-            {
-                // The <T> value in the matching JsonPropertyInfo<T> instance matches the parameter type.
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs:93]: This separation is no longer necessary -- the <T> value now always matches that of the supplied JsonPropertyInfo<T>.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs [hunk 4]
```diff
@@ -1129,25 +1121,6 @@ internal JsonPropertyDictionary<JsonPropertyInfo> CreatePropertyCache(int capaci
             return new JsonPropertyDictionary<JsonPropertyInfo>(Options.PropertyNameCaseInsensitive, capacity);
         }
 
-        private static JsonParameterInfo CreateConstructorParameter(
-            JsonParameterInfoValues parameterInfo,
-            JsonPropertyInfo jsonPropertyInfo,
-            bool sourceGenMode,
-            JsonSerializerOptions options)
-        {
-            if (jsonPropertyInfo.IsIgnored)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:1138]: Creating a placeholder is no longer necessary.

File: src/libraries/System.Text.Json/tests/Common/ConstructorTests/ConstructorTests.ParameterMatching.cs [hunk 5]
```diff
@@ -1219,7 +1219,7 @@ public async Task SmallObject_ClrDefaultParamValueUsed_WhenMatchingPropIgnored()
         {
             string json = @"{""Prop"":20}";
             var obj1 = await Serializer.DeserializeWrapper<SmallType_IgnoredProp_Bind_ParamWithDefaultValue>(json);
-            Assert.Equal(0, obj1.Prop);
+            Assert.Equal(5, obj1.Prop);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/ConstructorTests/ConstructorTests.ParameterMatching.cs:1222]: Updating the test in line with https://github.com/dotnet/runtime/issues/60082

File: src/libraries/System.Text.Json/tests/Common/TestClasses/TestClasses.Constructor.cs [hunk 6]
```diff
@@ -2057,7 +2057,7 @@ public class Point_MembersHave_JsonIgnore : ITestClass
         public void Verify()
         {
             Assert.Equal(0, X);
-            Assert.Equal(0, Y); // We don't set parameter default value here.
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/TestClasses/TestClasses.Constructor.cs:2060]: Updating in line with https://github.com/dotnet/runtime/issues/60082

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs (hunk 1 above)
Comment (krwq) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs:98]: I'd rather do opposite and not do it for large constructors but fine with me to leave this as is in this PR since there is already issue open

Comment (krwq) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs:98]: (also seems that small is more common case so I'd not want to reinforce that)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs [hunk 7]
```diff
@@ -1,42 +0,0 @@
...
-// The .NET Foundation licenses this file to you under the MIT license.
-
-using System.Diagnostics.CodeAnalysis;
-
-namespace System.Text.Json.Serialization.Metadata
-{
-    /// <summary>
-    /// Helper class used for calculating the default value for a given System.Type instance.
-    /// </summary>
-    internal sealed class DefaultValueHolder
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs:11]: ❤️ 🚀

**Comment by krwq** [comment]
given this scenario might affect benchmarks, I'd recommend to run perf tests, otherwise LGTM

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs [hunk 8]
```diff
@@ -11,108 +11,43 @@ namespace System.Text.Json.Serialization.Metadata
...
-        public virtual void Initialize(JsonParameterInfoValues parameterInfo, JsonPropertyInfo matchingProperty, JsonSerializerOptions options)
+        public JsonParameterInfo(JsonParameterInfoValues parameterInfoValues, JsonPropertyInfo matchingProperty)
         {
+            Debug.Assert(matchingProperty.IsConfigured);
+
             MatchingProperty = matchingProperty;
-            ClrInfo = parameterInfo;
-            Options = options;
-            ShouldDeserialize = true;
+            ShouldDeserialize = !matchingProperty.IsIgnored;
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs:43]: is ShouldDeserialize still used?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs [hunk 9]
```diff
@@ -11,108 +11,43 @@ namespace System.Text.Json.Serialization.Metadata
...
     {
-        private JsonTypeInfo? _jsonTypeInfo;
+        public JsonConverter EffectiveConverter => MatchingProperty.EffectiveConverter;
 
-        public JsonConverter ConverterBase { get; private set; } = null!;
+        // The default value of the parameter. This is `DefaultValue` of the `ParameterInfo`, if specified, or the `default` for the `ParameterType`.
+        public object? DefaultValue { get; private protected init; }
 
-        private protected bool MatchingPropertyCanBeNull { get; private set; }
+        public bool IgnoreNullTokensOnRead { get; }
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs:19]: make sure we have a test with null and IgnoreNullTokensOnRead making sure we use default value for that case (assuming that's even possible)

File: src/libraries/System.Text.Json/Common/ReflectionExtensions.cs [hunk 10]
```diff
@@ -319,14 +319,34 @@ public static bool IsKeyValuePair(this Type type, Type? keyValuePairType = null)
...
-            if (defaultValue == DBNull.Value && parameterInfo.ParameterType != typeof(DBNull))
+            if (defaultValue == DBNull.Value && parameterType != typeof(DBNull))
             {
                 return null;
             }
 
+#if !BUILDING_SOURCE_GENERATOR
+            // Default values of enums or nullable enums are represented using the underlying type and need to be cast explicitly
+            // cf. https://github.com/dotnet/runtime/issues/68647
+            if (parameterType.IsEnum)
```
Comment (krwq) [inline src/libraries/System.Text.Json/Common/ReflectionExtensions.cs:339]: this is ugly 😄 is there an issue against Type.DefaultValue about it? Why is that necessary?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/Common/ReflectionExtensions.cs:339]: Yep. It's unfortunate but necessary because the method is being used against the source generator's System.Reflection wrappers around Roslyn symbols. These checks are simply not applicable in that world.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs (hunk 7 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs:11]: My thoughts exactly :-)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs (hunk 8 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs:43]: It is, the converters are consulting it when deserializing constructor parameters.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs (hunk 9 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs:19]: Seems that don't test for that. I've added a test even though it's not in scope for this change.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs:98]: Fundamentally this is what causing the bug in https://github.com/dotnet/runtime/issues/60082 so I think we should fix it.

File: src/libraries/System.Text.Json/tests/Common/ConstructorTests/ConstructorTests.ParameterMatching.cs [hunk 11]
```diff
@@ -361,6 +361,14 @@ public async Task IgnoreNullValues_DontSetNull_ToConstructorArguments_ThatCantBe
...
 
         [Fact]
+        public async Task IgnoreNullValues_SetDefaultConstructorParameter_ToConstructorArguments_ThatCanBeNull()
+        {
+            var options = new JsonSerializerOptions { IgnoreNullValues = true };
+            NullArgTester result = await Serializer.DeserializeWrapper<NullArgTester>(@"{""String"":null}", options);
+            Assert.Equal("defaultStr", result.String);
+        }
+
+            [Fact]
```
Comment (krwq) [inline src/libraries/System.Text.Json/tests/Common/ConstructorTests/ConstructorTests.ParameterMatching.cs:371]: nit: indent

**Review by krwq — APPROVED** [review]
LGTM
